reference/test/preconditioner/ilu.cpp

```diff
@@ -0,0 +1,317 @@
+};
+
+
+TEST_F(Ilu, KnowsItsExecutor)
+{
+    auto ilu_factory = ilu_prec_type::build().on(exec);
+
+    ASSERT_EQ(ilu_factory->get_executor(), exec);
+}
+
+
+TEST_F(Ilu, CanSetLSolverFactory)
+{
+    auto ilu_factory =
+        ilu_prec_type::build().with_l_solver_factory(l_factory).on(exec);
+
+    ASSERT_EQ(ilu_factory->get_parameters().l_solver_factory, l_factory);
+}
+
+
+TEST_F(Ilu, CanSetUSolverFactory)
+{
+    auto ilu_factory =
+        ilu_prec_type::build().with_u_solver_factory(u_factory).on(exec);
+
+    ASSERT_EQ(ilu_factory->get_parameters().u_solver_factory, u_factory);
+}
+
+
+TEST_F(Ilu, BuildsDefaultWithoutThrowing)
+{
+    auto ilu_pre_default_factory = ilu_prec_type::build().on(exec);
+    auto par_ilu_fact =
+        gko::factorization::ParIlu<value_type>::build().on(exec);
+    auto par_ilu = par_ilu_fact->generate(mtx);
+
+    ASSERT_NO_THROW(ilu_pre_default_factory->generate(gko::share(par_ilu)));
+}
+
+
+TEST_F(Ilu, BuildsCustomWithoutThrowing)
+{
+    auto par_ilu_fact =
+        gko::factorization::ParIlu<value_type>::build().on(exec);
+    auto par_ilu = par_ilu_fact->generate(mtx);
+
+    ASSERT_NO_THROW(ilu_pre_factory->generate(gko::share(par_ilu)));
+}
+
+
+TEST_F(Ilu, ThrowOnWrongInput)
+{
+    ASSERT_THROW(ilu_pre_factory->generate(l_factor), gko::NotSupported);
+}
+
+
+TEST_F(Ilu, SolvesDefaultSingleRhsWithParIlu)
+{
+    const auto b = gko::initialize<Mtx>({1.0, 3.0, 6.0}, exec);
+    auto x = Mtx::create(exec, gko::dim<2>{3, 1});
+    auto par_ilu_fact =
+        gko::factorization::ParIlu<value_type>::build().on(exec);
+    auto par_ilu = par_ilu_fact->generate(mtx);
+
+    auto preconditioner =
+        default_ilu_prec_type::build().on(exec)->generate(gko::share(par_ilu));
+    preconditioner->apply(b.get(), x.get());
+
+    // Since it is a preconditioner, the default solve is not accurate
+    // Applies only to BiCGSTAB   TODO: Replace with 1e-14 when TRS is default
+    GKO_ASSERT_MTX_NEAR(x.get(), l({-0.125, 0.25, 1.0}), 5e-1);
+}
+
+
+TEST_F(Ilu, SolvesSingleRhsWithParIlu)
+{
+    const auto b = gko::initialize<Mtx>({1.0, 3.0, 6.0}, exec);
+    auto x = Mtx::create(exec, gko::dim<2>{3, 1});
+    auto par_ilu_fact =
+        gko::factorization::ParIlu<value_type>::build().on(exec);
+    auto par_ilu = par_ilu_fact->generate(mtx);
+
+    auto preconditioner = ilu_pre_factory->generate(gko::share(par_ilu));
+    preconditioner->apply(b.get(), x.get());
+
+    GKO_ASSERT_MTX_NEAR(x.get(), l({-0.125, 0.25, 1.0}), 1e-14);
+}
+
+
+TEST_F(Ilu, SolvesSingleRhsWithComposition)
+{
+    const auto b = gko::initialize<Mtx>({1.0, 3.0, 6.0}, exec);
+    auto x = Mtx::create(exec, gko::dim<2>{3, 1});
+    auto composition = gko::Composition<value_type>::create(l_factor, u_factor);
+
+    auto preconditioner = ilu_pre_factory->generate(gko::share(composition));
+    preconditioner->apply(b.get(), x.get());
+
+    GKO_ASSERT_MTX_NEAR(x.get(), l({-0.125, 0.25, 1.0}), 1e-14);
+}
+
+
+TEST_F(Ilu, SolvesSingleRhsWithLU)
```

Comment:
could it be `AppliesToVectorWithComposition`?
other tests for apply use this sentence.

Reply:
Yes I could rename it from `Solves` to `Applies`, but I thought I go with the more specific description here.

The `ToVectorWithComposition` would not be accurate enough because I want to specify that it is a single vector, so a single right hand side. Also, in this test, it is with two separate matrices (L and U), the Composition would be the test above.

We could standardize it, but I don't think that there would be any real benefit from it. Currently, these names only help the developer to see which part does not work, so a descriptive name is very important. But apart from that, we don't have any standards / requirements for the test name as far as I know.

Reply:
The name is good if we do not have the standard of this.